Implement trans for the MIR Switch terminator

wesleywiser opened the pull request:
Fixes #29574

rust-highfive commented:
r? @arielb1

(rust_highfive has picked a reviewer for you, use r? to override)

wesleywiser commented:
r? @nikomatsakis

nagisa commented on src/test/run-pass/mir_trans_switch.rs:
```diff
@@ -0,0 +1,35 @@
+        Abc::C => 3,
+        Abc::D => 4,
+        Abc::B(_) => 2,
+        Abc::A(_) => 1,
```
You need at least a test for the `_ =>` case.

nagisa commented on src/librustc_trans/trans/mir/block.rs:
```diff
@@ -46,8 +47,23 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+
+                //Use the diverge block for the default case since all enum variants have targets
+                //and it shouldn't ever be called
+                let diverge_llbb = self.llblock(mir::DIVERGE_BLOCK);
```
What about the regular default case, i.e. `_ =>`

nagisa commented on src/librustc_trans/trans/mir/block.rs:
```diff
@@ -46,8 +47,23 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+                let discr = adt::trans_get_discr(bcx, &*represented_ty, discr_lvalue.llval, None);
+
+                //Use the diverge block for the default case since all enum variants have targets
+                //and it shouldn't ever be called
```
NIT: a space after `//`

dotdash commented on src/librustc_trans/trans/mir/block.rs:
```diff
@@ -46,8 +47,23 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+                let discr_lvalue = self.trans_lvalue(bcx, discr);
+                let discr = adt::trans_get_discr(bcx, &*represented_ty, discr_lvalue.llval, None);
+
+                //Use the diverge block for the default case since all enum variants have targets
```
For exhaustive matches, the default arm of the switch statement should go to a block that only contains an `unreachable` instruction. LLVM uses that for optimizations.

wesleywiser commented:
@nagisa Fixed nit and added test case for `_ => `. It looks like the default case is translated to all other variants pointing to the same block. (You can see the graphviz for `foo2` here: https://gist.github.com/wesleywiser/698dbdd7945ec51a1e3c)

@dotdash I think I fixed this in my last commit. Does that look right to you?

nagisa commented:
@wesleywiser @dotdash regarding `unreachable`…

`DIVERGE_BLOCK` is guaranteed to only contain the `Diverge` terminator (and no statements), which translates down straight to `unreachable`. That being said, @wesleywiser, I would remove the b834929

Re `_=>`, I see, its all good then.

dotdash commented:
@nagisa AIUI, the fact that the diverge block can contain an `unreachable` is the hack. The block is meant for unwinding/panicking, but we always translate it, even if there's no way to actually reach it , and only in that case we put the `unreachable` there. In a perfect world, we wouldn't even create the block in that case.

nagisa commented:
@dotdash but all you can really do in that block, really, is having a `resume`, since landing block itself will be dependent on the location where function was generated, no?

I’d rather have `DIVERGE_BLOCK` be unconditionally unreachable (and translated lazily, only if necessary/removed by CFG simplification), for cases like this (something to branch into after call to diverging function comes into mind as similar case) and a separate Resume terminator which is emitted by `librustc_mir/build/scope.rs` after all necessary cleanups.

nikomatsakis commented:
So, it seems to me that perhaps we should just remove `DIVERGE_BLOCK` altogether, and instead replicate the `DIVERGE` terminator at the end of every cleanup sequence. What @nagisa says is, I think, true -- that is, that there isn't much useful code there. I'm not sure I included it exactly, probably just by analogy to the "return" block. Certainly, one could also make the argument that RETURN is not needed. I think mainly I was thinking that it's often easier for analysis purposes if you have a fixed set of "postdominator" blocks, but that may not be really true.

nikomatsakis commented:
As far as having a special block for unreachable code, that seems fine, but it seems like something specific to trans, and not necessarily something that needs to be made manifest in the MIR.

nikomatsakis commented on src/librustc_trans/trans/mir/block.rs:
```diff
@@ -46,8 +47,25 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
-                unimplemented!()
+            mir::Terminator::Switch { ref discr, ref adt_def, ref targets } => {
+                let adt_ty = bcx.tcx().lookup_item_type(adt_def.did).ty;
+                let represented_ty = adt::represent_type(bcx.ccx(), adt_ty).clone();
```
Nit: I don't think you need to call `clone()` here, do you?

nikomatsakis commented on src/librustc_trans/trans/mir/block.rs:
```diff
@@ -46,8 +47,25 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+                let represented_ty = adt::represent_type(bcx.ccx(), adt_ty).clone();
+
+                let discr_lvalue = self.trans_lvalue(bcx, discr);
+                let discr = adt::trans_get_discr(bcx, &*represented_ty, discr_lvalue.llval, None);
```
Nit: `&represented_ty` should be fine, not `&*represented_ty`

nikomatsakis commented on src/librustc_trans/trans/mir/block.rs:
```diff
@@ -46,8 +47,25 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+                // The else branch of the Switch can't be hit, so branch to an unreachable
+                // instruction so LLVM knows that
+                let unreachable_blk = bcx.fcx.new_temp_block("enum-variant-unreachable");
+                build::Unreachable(unreachable_blk);
```
As @nagisa says, it might be nice to have just one such block (created lazilly), we could store it in the "MIR trans" state.

nikomatsakis commented on src/librustc_trans/trans/mir/block.rs:
```diff
@@ -46,8 +47,25 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+                build::Unreachable(unreachable_blk);
+
+                let switch = build::Switch(bcx, discr, unreachable_blk.llbb, targets.len());
+                for (adt_variant, target) in adt_def.variants.iter().zip(targets) {
```
Nit: Please add an `assert_eq!(adt_def.variants.len(), targets.len())`

wesleywiser commented:
@nikomatsakis Fixed nits. 

> it might be nice to have just one such block (created lazilly), we could store it in the "MIR trans" state.

Do you want me to work on that in this PR?

nagisa commented:
You can leave it as it is currently, however, if you can add a FIXME of some sort to that effect.

wesleywiser commented:
Ok. Should I tidy the branch up too while I'm in there?

nagisa commented:
Sure!

wesleywiser commented:
I've squashed all my changes and rebased against master.

nikomatsakis commented:
@bors r+

nice!

bors commented:
:pushpin: Commit 2b15361 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 2b1536129827dbad754388695558ac37de063941 -->

bors commented:
:hourglass: Testing commit 2b15361 with merge 073b0f9...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7437), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7455), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/711), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7393), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7413), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7397), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/623), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2576), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7467), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7458), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7424), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2424), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2431), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2434), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2421), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1677), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2415)
